cpp/src/unary/math_ops.cu

```diff
@@ -229,89 +227,239 @@ struct DeviceNot {
+  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
+                                           rmm::mr::device_memory_resource* mr,
+                                           cudaStream_t stream)
+  {
+    if (input.is_empty()) return empty_like(input);
+    auto dictionary_col = dictionary_column_view(input);
+    return type_dispatcher(
+      dictionary_col.keys().type(), dictionary_dispatch{}, dictionary_col, mr, stream);
   }
 
-  template <typename T, typename std::enable_if_t<!std::is_arithmetic<T>::value>* = nullptr>
+  template <typename T,
+            typename std::enable_if_t<!std::is_arithmetic<T>::value and
+                                      !std::is_same<T, dictionary32>::value>* = nullptr>
   std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
-                                           cudf::unary_op op,
                                            rmm::mr::device_memory_resource* mr,
                                            cudaStream_t stream)
   {
-    CUDF_FAIL("Unsupported datatype for operation");
+    CUDF_FAIL("Unsupported data type for operation");
   }
 };
 
-template <typename F>
+template <typename UFN>
 struct BitwiseOpDispatcher {
   template <typename T, typename std::enable_if_t<std::is_integral<T>::value>* = nullptr>
   std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
-                                           cudf::unary_op op,
                                            rmm::mr::device_memory_resource* mr,
                                            cudaStream_t stream)
   {
-    return launch<T, F>(input, op, mr, stream);
+    return transform_fn<T, UFN>(input.begin<T>(),
+                                input.end<T>(),
+                                copy_bitmask(input, stream, mr),
+                                input.null_count(),
+                                mr,
+                                stream);
+  }
+
+  struct dictionary_dispatch {
+    template <typename T, typename std::enable_if_t<std::is_integral<T>::value>* = nullptr>
+    std::unique_ptr<cudf::column> operator()(cudf::dictionary_column_view const& input,
+                                             rmm::mr::device_memory_resource* mr,
+                                             cudaStream_t stream)
+    {
+      return transform_fn<T, UFN>(input, mr, stream);
+    }
+
+    template <typename T, typename std::enable_if_t<!std::is_integral<T>::value>* = nullptr>
+    std::unique_ptr<cudf::column> operator()(cudf::dictionary_column_view const& input,
+                                             rmm::mr::device_memory_resource* mr,
+                                             cudaStream_t stream)
+    {
+      CUDF_FAIL("dictionary keys type not supported for this operation");
+    }
+  };
+
+  template <typename T,
+            typename std::enable_if_t<!std::is_integral<T>::value and
```

Comment:
Does this need to be here?  I'm guessing a dictionary doesn't check out as `is_integral`

Reply:
Yes, dictionary is not integral.